packages/wdio-allure-reporter/src/index.ts

```diff
@@ -1,605 +1,686 @@
+            const suite = this.allure.getCurrentSuite()
+            if (!suite || !(suite.currentStep instanceof Step)) {
+                return
+            }
+
+            this.allure.endStep('passed')
+        }
+    }
+
+    onHookStart(hook: HookStats) {
+        // ignore global hooks
+        if (!hook.parent || !this.allure.getCurrentSuite()) {
+            return false
+        }
+
+        // add beforeEach / afterEach hook as step to test
+        if (this._options.disableMochaHooks && isMochaEachHooks(hook.title)) {
+            if (this.allure.getCurrentTest()) {
+                this.allure.startStep(hook.title)
+            }
+            return
+        }
+
+        // don't add hook as test to suite for mocha All hooks
+        if (this._options.disableMochaHooks && isMochaAllHooks(hook.title)) {
+            return
+        }
+
+        // add hook as test to suite
+        this.onTestStart(hook)
+    }
+
+    onHookEnd(hook: HookStats) {
+        // ignore global hooks
+        if (!hook.parent || !this.allure.getCurrentSuite() || (this._options.disableMochaHooks && !isMochaAllHooks(hook.title) && !this.allure.getCurrentTest())) {
+            return false
+        }
+
+        // set beforeEach / afterEach hook (step) status
+        if (this._options.disableMochaHooks && isMochaEachHooks(hook.title)) {
+            if (hook.error) {
+                this.allure.endStep('failed')
+            } else {
+                this.allure.endStep('passed')
+            }
+            return
+        }
+
+        // set hook (test) status
+        if (hook.error) {
+            if (this._options.disableMochaHooks && isMochaAllHooks(hook.title)) {
+                this.onTestStart(hook)
+                this.attachScreenshot()
+            }
+            this.onTestFail(hook)
+        } else if (this._options.disableMochaHooks || this._options.useCucumberStepReporter) {
+            if (!isMochaAllHooks(hook.title)) {
+                this.onTestPass()
+
+                // remove hook from suite if it has no steps
+                if (this.allure.getCurrentTest().steps.length === 0 && !this._options.useCucumberStepReporter) {
+                    this.allure.getCurrentSuite().testcases.pop()
+                } else if (this._options.useCucumberStepReporter) {
+                    // remove hook when it's registered as a step and if it's passed
+                    const step = this.allure.getCurrentTest().steps.pop()
+
+                    // if it had any attachments, reattach them to current test
+                    if (step && step.attachments.length >= 1) {
+                        step.attachments.forEach(attachment => {
+                            this.allure.getCurrentTest().addAttachment(attachment)
+                        })
+                    }
+                }
+            }
+        } else if (!this._options.disableMochaHooks) this.onTestPass()
+    }
+
+    addLabel({
+        name,
+        value
+    }: any) {

```

Comment:
can we type these and the following?

Reply:
Personally, I don't think it's worth it. These are event listeners, adding an interface would bring no type-checking value because process.emit and process.on will always use anonyous type. Regardless, I think these event handlers should be private

Reply:
The way it is used in unit test also make it really troublesome to define all those typings, with no real benefits
```
    it('should do nothing if no tests run', () => {
        const reporter = new AllureReporter()
        expect(reporter.addLabel({})).toEqual(false)
        expect(reporter.addStory({})).toEqual(false)
        expect(reporter.addFeature({})).toEqual(false)
        expect(reporter.addSeverity({})).toEqual(false)
        expect(reporter.addIssue({})).toEqual(false)
        expect(reporter.addTestId({})).toEqual(false)
        expect(reporter.addEnvironment({})).toEqual(false)
        expect(reporter.addDescription({})).toEqual(false)
        expect(reporter.addAttachment({})).toEqual(false)
        expect(reporter.startStep('test')).toEqual(false)
        expect(reporter.endStep('passed')).toEqual(false)
        expect(reporter.addStep({})).toEqual(false)
        expect(reporter.addArgument({})).toEqual(false)
    })
   ```